dev-support/ci/selective_ci_checks.sh

```diff
@@ -0,0 +1,470 @@
+        "src/..../java"
+    )
+    show_changed_files
+
+    if [[ $(count_changed_files) != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} author"
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_bats() {
+    start_end::group_start "Check if bats is needed"
+    local pattern_array=(
+        "\.bash$"
+        "\.bats$"
+        "\.sh$" # includes hadoop-ozone/dev-support/checks/bats.sh
+    )
+    show_changed_files
+
+    if [[ $(count_changed_files) != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} bats"
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_checkstyle() {
+    start_end::group_start "Check if checkstyle is needed"
+    local pattern_array=(
+        "^hadoop-ozone/dev-support/checks/checkstyle.sh"
+        "^hadoop-hdds/dev-support/checkstyle"
+        "pom.xml"
+        "src/..../java"
+    )
+    show_changed_files
+
+    if [[ $(count_changed_files) != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} checkstyle"
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_docs() {
+    start_end::group_start "Check if docs is needed"
+
+    if [[ ${COUNT_DOC_CHANGED_FILES} != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} docs"
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_dependency() {
+    start_end::group_start "Check if dependency is needed"
+    local pattern_array=(
+        "^hadoop-ozone/dev-support/checks/dependency.sh"
+        "^hadoop-ozone/dist/src/main/license/update-jar-report.sh"
+        "^hadoop-ozone/dist/src/main/license/jar-report.txt"
+        "pom.xml"
+    )
+    show_changed_files
+
+    dependency_check_needed=false
+    if [[ $(count_changed_files) != "0" ]]; then
+        dependency_check_needed=true
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_findbugs() {
+    start_end::group_start "Check if findbugs is needed"
+    local pattern_array=(
+        "^hadoop-ozone/dev-support/checks/findbugs.sh"
+        "findbugsExcludeFile.xml"
+        "pom.xml"
+        "src/..../java"
+    )
+    show_changed_files
+
+    if [[ $(count_changed_files) != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} findbugs"
+    fi
+
+    start_end::group_end
+}
+
+function check_needs_unit_test() {
+    start_end::group_start "Check if unit test is needed"
+    local pattern_array=(
+        "^hadoop-ozone/dev-support/checks/_mvn_unit_report.sh"
+        "^hadoop-ozone/dev-support/checks/unit.sh"
+        "pom.xml"
+        "src/..../java"
+        "src/..../resources"
+    )
+    show_changed_files
+
+    if [[ $(count_changed_files) != "0" ]]; then
+        BASIC_CHECKS="${BASIC_CHECKS} unit"
+    fi
+
+    start_end::group_end
+}
+
+function calculate_test_types_to_run() {
+    start_end::group_start "Count core/other files"
+    COUNT_CORE_OTHER_CHANGED_FILES=$((COUNT_ALL_CHANGED_FILES - COUNT_COMPOSE_CHANGED_FILES - COUNT_DOC_CHANGED_FILES - COUNT_JUNIT_CHANGED_FILES - COUNT_KUBERNETES_CHANGED_FILES - COUNT_ROBOT_CHANGED_FILES))
```

Comment:
I'm a little concerned that this calculation will silently go wrong if someone updates the regexes so they overlap.

is that worth worrying about?  we could keep a running list matched filenames, so we can generate an accurate match count.  Something like this:

https://github.com/adoroszlai/hadoop-ozone/compare/57d10fe34b396866b5f4b92e4ed9d62bcaf7e7a2...GeorgeJahad:5000-with-list-of-matches

Reply:
I'm also concerned about this.  Will try to adopt your improvement.  Also considering adding bats test, but the script needs to be restructured a bit for that.

Reply:
thanks, I didn't necessarily mean we should do it that way.  I just wanted to show that it doesn't seem too hard to make it safer.  I was also thinking about trying to use an associative array as a set.

Reply:
I have merged your changes into my branch, then went on to unify code that was now repeated in 3 functions.  Thanks a lot!

Reply:
Your bats changes look interesting too.  I'll take a look when I can.  Hopefully today.